lib/internal/fs/recursive_watch.js

```diff
@@ -0,0 +1,288 @@
+'use strict';
+
+const {
+  ArrayPrototypePush,
+  SafePromiseAll,
+  Promise,
+  PromisePrototypeThen,
+  SafeMap,
+  StringPrototypeStartsWith,
+  StringPrototypeSlice,
+  StringPrototypeLastIndexOf,
+  SymbolAsyncIterator,
+} = primordials;
+
+const { EventEmitter } = require('events');
+const assert = require('internal/assert');
+const {
+  AbortError,
+  codes: {
+    ERR_INVALID_ARG_VALUE,
+  },
+} = require('internal/errors');
+const { getValidatedPath } = require('internal/fs/utils');
+const { kFSWatchStart, StatWatcher } = require('internal/fs/watchers');
+const { kEmptyObject } = require('internal/util');
+const { validateBoolean, validateAbortSignal } = require('internal/validators');
+const path = require('path');
+
+let internalSync;
+let internalPromises;
+
+function lazyLoadFsPromises() {
+  internalPromises ??= require('fs/promises');
+  return internalPromises;
+}
+
+function lazyLoadFsSync() {
+  internalSync ??= require('fs');
+  return internalSync;
+}
+
+async function traverse(dir, files = new SafeMap()) {
+  const { opendir } = lazyLoadFsPromises();
+
+  const filenames = await opendir(dir);
+  const subdirectories = [];
+
+  for await (const file of filenames) {
+    const f = path.join(dir, file.name);
+
+    files.set(f, file);
+
+    if (file.isDirectory()) {
+      ArrayPrototypePush(subdirectories, traverse(f, files));
+    }
+  }
+
+  await SafePromiseAll(subdirectories);
```

Comment:
Now that #45175 has landed:
```suggestion
  await SafePromiseAllReturnVoid(subdirectories);
```